### ISSUE TITLE
[Feature] Add Gradient Clipping

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ProjectOdyssey#2666]
## Problem

Large gradients can cause training instability (exploding gradients), especially in RNNs and deep networks.

### 1. Clip by Value

```mojo
# shared/training/gradient_clipping.mojo
fn clip_grad_value(parameters: List[ExTensor], clip_value: Float32):
    """Clip gradients by absolute value."""
    for param in parameters:
        if not param.grad:
            continue
        
        var grad = param.grad.value()
        for i in range(grad.numel()):
            if grad[i] > clip_value:
                grad[i] = clip_value
            elif grad[i] < -clip_value:
                grad[i] = -clip_value

# Usage
fn train_step(model: Model, optimizer: Optimizer):
    var loss = forward_and_loss(model)
    loss.backward()
    
    # Clip gradients to [-1.0, 1.0]
    clip_grad_value(model.parameters(), 1.0)
    
    optimizer.step()
```

### 2. Clip by Norm (More Common)

```mojo
fn clip_grad_norm(parameters: List[ExTensor], max_norm: Float32) -> Float32:
    """Clip gradients by global norm.
    
    Returns:
        Total gradient norm before clipping.
    """
    # Compute total gradient norm
    var total_norm: Float32 = 0.0
    for param in parameters:
        if not param.grad:
            continue
        
        var grad = param.grad.value()
        for i in range(grad.numel()):
            total_norm += grad[i] * grad[i]
    
    total_norm = math.sqrt(total_norm)
    
    # Clip if needed
    if total_norm > max_norm:
        var clip_coef = max_norm / (total_norm + 1e-6)
        
        for param in parameters:
            if not param.grad:
                continue
            
            var grad = param.grad.value()
            for i in range(grad.numel()):
                grad[i] *= clip_coef
    
    return total_norm

# Usage
fn train_step(model: Model, optimizer: Optimizer):
    var loss = forward_and_loss(model)
    loss.backward()
    
    # Clip gradient norm to max 1.0
    var grad_norm = clip_grad_norm(model.parameters(), max_norm=1.0)
    
    if grad_norm > 10.0:
        print("Warning: Large gradient norm:", grad_norm)
    
    optimizer.step()
```

### 3. Per-Parameter Norm Clipping

```mojo
fn clip_grad_norm_per_param(parameters: List[ExTensor], max_norm: Float32):
    """Clip each parameter's gradient independently."""
    for param in parameters:
        if not param.grad:
            continue
        
        var grad = param.grad.value()
        
        # Compute this parameter's gradient norm
        var param_norm: Float32 = 0.0
        for i in range(grad.numel()):
            param_norm += grad[i] * grad[i]
        param_norm = math.sqrt(param_norm)
        
        # Clip if needed
        if param_norm > max_norm:
            var clip_coef = max_norm / (param_norm + 1e-6)
            for i in range(grad.numel()):
                grad[i] *= clip_coef
```

### 4. Adaptive Clipping

```mojo
struct AdaptiveGradientClipper:
    """Adaptively adjust clipping threshold based on gradient history."""
    var history: List[Float32]
    var history_size: Int
    var percentile: Float32
    
    fn __init__(out self, history_size: Int = 100, percentile: Float32 = 95.0):
        self.history = List[Float32]()
        self.history_size = history_size
        self.percentile = percentile
    
    fn clip(mut self, parameters: List[ExTensor]) -> Float32:
        """Clip using adaptive threshold."""
        # Compute current gradient norm
        var grad_norm = compute_grad_norm(parameters)
        
        # Update history
        self.history.append(grad_norm)
        if len(self.history) > self.history_size:
            self.history.pop(0)
        
        # Compute adaptive threshold (e.g., 95th percentile)
        var threshold = compute_percentile(self.history, self.percentile)
        
        # Clip using threshold
        if grad_norm > threshold:
            clip_grad_norm(parameters, threshold)
        
        return grad_norm

# Usage
var clipper = AdaptiveGradientClipper(history_size=100, percentile=95.0)
for batch in dataloader:
    var loss = forward_and_loss(model, batch)
    loss.backward()
    
    var grad_norm = clipper.clip(model.parameters())
    optimizer.step()
```

### 5. Integration with Optimizer

```mojo
struct OptimizerWithClipping:
    var base_optimizer: Optimizer
    var max_grad_norm: Float32
    
    fn step(mut self, parameters: List[ExTensor]):
        """Step with automatic gradient clipping."""
        var grad_norm = clip_grad_norm(parameters, self.max_grad_norm)
        self.base_optimizer.step(parameters)
        return grad_norm

# Usage
var optimizer = OptimizerWithClipping(
    base_optimizer=Adam(lr=0.001),
    max_grad_norm=1.0
)

for batch in dataloader:
    var loss = forward_and_loss(model, batch)
    loss.backward()
    var grad_norm = optimizer.step(model.parameters())
```

### 6. Monitoring Gradients

```mojo
fn monitor_gradients(parameters: List[ExTensor]) -> Dict[String, Float32]:
    """Compute gradient statistics for monitoring."""
    var stats = Dict[String, Float32]()
    
    var total_norm: Float32 = 0.0
    var max_grad: Float32 = 0.0
    var min_grad: Float32 = Float32.infinity()
    
    for param in parameters:
        if not param.grad:
            continue
        
        var grad = param.grad.value()
        for i in range(grad.numel()):
            var abs_grad = math.abs(grad[i])
            total_norm += grad[i] * grad[i]
            max_grad = math.max(max_grad, abs_grad)
            min_grad = math.min(min_grad, abs_grad)
    
    stats["grad_norm"] = math.sqrt(total_norm)
    stats["grad_max"] = max_grad
    stats["grad_min"] = min_grad
    
    return stats

# Usage in training loop
var loss = forward_and_loss(model)
loss.backward()

var grad_stats = monitor_gradients(model.parameters())
print("Grad norm:", grad_stats["grad_norm"])
print("Grad max:", grad_stats["grad_max"])

clip_grad_norm(model.parameters(), max_norm=1.0)
optimizer.step()
```

## When to Use

- **RNNs/LSTMs**: Prevent exploding gradients
- **Large learning rates**: Stabilize early training
- **Deep networks**: Prevent gradient explosion
- **Unstable training**: Loss becomes NaN

## Typical Values

- **Conservative**: max_norm = 0.5
- **Standard**: max_norm = 1.0
- **Aggressive**: max_norm = 5.0

## Benefits

- Prevent exploding gradients
- Stabilize training
- Enable higher learning rates
- Standard practice for RNNs

## Impact

- **Severity**: LOW
- **Stability**: More stable training
- **Convergence**: Better for RNNs and deep networks
